### PR TITLE
Add support for struct type in ORC writer

### LINKED ISSUE #7830: [FEA] Support struct types in ORC writer
Struct support:

- [x] Reader (https://github.com/rapidsai/cudf/pull/8599)
- [ ] Writer

### PR DESCRIPTION
**Pull request by vuule**
Fixes #7830, #8443

Features:
- Use the new table metadata type that matches the table hierarchy, `table_input_metadata`.
- Support struct columns in the writer.

Changes:
- Null masks are encoded as aligned rowgroups to avoid invalid bits when the number of encoded rows is not divisible by 8 (except for the last rowgroup in each stripe). This also affects list columns. The issue is equivalent to https://github.com/rapidsai/cudf/issues/6763 (boolean columns only).
- Added pushdown masks that are used to determine which child elements should not be encoded, including null mask bits.
- Use pushdown masks for rowgroup alignment, null mask encoding and value encoding.
- Separated the null mask encoding from value encoding - can be further moved to a separate kernel call.

Breaking because the table metadata type has changed.

### CODE REVIEW

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Review by revans2 — APPROVED** [review]
From the java side this look good.

**Comment by jlowe** [comment]
Note that this does not add Java bindings for the new struct writing support.  Adding the Java bindings is being tracked by #9233.

**Comment by vuule** [comment]
rerun tests

File: python/cudf/cudf/_lib/orc.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/orc.pyx:297]: In `write_orc` the corresponding block checks for any non-RangeIndex type. Is that discrepancy intentional or meaningful? Or do you always get either a MultiIndex or a RangeIndex?

File: python/cudf/cudf/tests/test_orc.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_orc.py:1385]: Is there a reason that `list_struct_buff` is a module-scope variable rather than a fixture? If it's only ever read, could it be a module-scope fixture?

**Review by vyasr — APPROVED** [review]
Python/Cython approval, with a couple of small questions.

C++/CUDA looked OK, but probably best to get approval from someone with greater familiarity. I didn't review those files in depth.

Is there a reason that submodule updates are present in this PR?

**Comment by vuule** [comment]
> Is there a reason that submodule updates are present in this PR? The current build failure may be related since it looked like it was failing to pull in jitify.

They got pulled in when I merged 21.10. Not sure what I did wrong and how to revert it. I have these two files marked as modified in git. When I try to update submodules, it fails with `No url found for submodule path 'thirdparty/jitify' in .gitmodules`.

**Comment by vuule** [comment]
rerun tests

File: python/cudf/cudf/_lib/orc.pyx
Comment (vuule) [reply python/cudf/cudf/_lib/orc.pyx:297]: I have to defer to @devavret or @galipremsagar on this one. The Python changes in this PR are based on the equivalent Parquet PR https://github.com/rapidsai/cudf/pull/7461.

Comment (devavret) [reply python/cudf/cudf/_lib/orc.pyx:297]: This code block is a replacement of the logic of `utils.get_column_names` which assigns the names of the columns to metadata. We don't want to update the metadata for RangeIndex because metadata contains only information on libcudf column_views. We could get a single series materialized index too, which is captured by the second part of the branch.

Regarding writing RangeIndex, this is currently broken #7011 (and re-reported in #9216) because it might take some smarts to merge range indices across different calls to `write_table`. The RangeIndex only gets written in the pandas specific metadata and passes straight through libcudf untouched. If you have some time, I'd invite you to take on #7011

**Review by devavret — COMMENTED** [review]
Partial review but GitHub won't let me reply to @vyas' comment without publishing my review. 🤷‍♂️

File: cpp/include/cudf/io/types.hpp
Comment (vuule) [reply cpp/include/cudf/io/types.hpp:217]: it's not for ORC :) I'll remove the comment since this struct is supposed to be format agnostic now.

**Review by devavret — APPROVED** [review]
LGTM

File: cpp/src/io/orc/writer_impl.cu
Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:209]: I'm actually finding such APIs harder to use, since I have to include the condition in the calling code. Returning an optional allows to use value_or and get more concise code. I'll check if this is actually the case for the use of this function and split the API if it's not.

Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:209]: welp, it's not the case, returning optional here is useless. Breaking up the API as suggested :)

File: cpp/src/io/orc/stripe_init.cu
Comment (vuule) [reply cpp/src/io/orc/stripe_init.cu:567]: The doc is in the header. Some existing functions have duplicated docs.

**Review by rgsl888prabhu — COMMENTED** [review]
Initial pass, still going through PR.

File: cpp/src/io/orc/stripe_enc.cu
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:626]: not sure if I understand the question, hopefully this answers it:
If the current (child) column does not have a null mask, we don't need to encode the present stream even if the parent has a pushdown mask. 
If the current (child) column also has children, we don't need to do anything here since the pushdown masks are recomputed before the encode and the grandparent pushdown mask will be reused and passed down.

File: cpp/src/io/orc/stripe_enc.cu
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:633]: I don't think it can.
simplified relevant part of the loop:
```
buf_available_bits = encode_block_size * 8 - numvals
nrows = buf_available_bits // at most
numvals += nrows // at most

numvals -= num_encoded // can be zero
```
`nrows + numvals is always <= encode_block_size * 8` always holds

Comment (rgsl888prabhu) [reply cpp/src/io/orc/stripe_enc.cu:633]: Ok, missed that part

**Review by rgsl888prabhu — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**
